cmd/staterecovery/staterecovery.go

```diff
@@ -0,0 +1,85 @@
+package staterecovery
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/trie"
+	"github.com/ethereum/go-ethereum/trie/triedb/hashdb"
+)
+
+func RecreateMissingStates(chainDb ethdb.Database, bc *core.BlockChain, cacheConfig *core.CacheConfig) error {
+	start := time.Now()
+	current := bc.Genesis().NumberU64() + 1
+	genesisBlock := bc.GetBlockByNumber(current - 1)
+	if genesisBlock == nil {
+		return errors.New("genesis block is missing")
+	}
+	// find last available block - we cannot rely on bc.CurrentBlock()
+	last := current
+	for bc.GetBlockByNumber(last) != nil {
+		last++
```

Comment:
Instead of this let's just break when calling GetBlockByNumber below if it returns nil

Reply:
Basically, this might take a while to loop through and while it's happening the user gets no feedback on how far into the process is. For the "target" value in the log, we can log the blockchain's latest header number maybe, even if it's inaccurate.